### ISSUE TITLE
spirv-opt: New crash at glslang known_good spirv-opt with previously working SPIR-V shader

### ISSUE THREAD

**Issue by danginsburg** [KhronosGroup/SPIRV-Tools#903]
I am using spirv-tools to commit 99cd25c4139e0dc914ab8a5a3b75e6fed0ad1329 marked in known_good.json in glslang repository.  

The attached shader crashes in spirv-opt with the following callstack:

```
>	vfx_gl.dll!spvtools::opt::AggressiveDCEPass::AggressiveDCE(class spvtools::ir::Function *)	C++
 	vfx_gl.dll!spvtools::opt::Pass::ProcessCallTreeFromRoots(class std::function<bool > &,class std::unordered_map<unsigned int,class spvtools::ir::Function *,struct std::hash<unsigned int>,struct std::equal_to<unsigned int>,class std::allocator<struct std::pair<unsigned int const ,class spvtools::ir::Function *> > > const &,class std::queue<unsigned int,class std::deque<unsigned int,class std::allocator<unsigned int> > > *)	C++
 	vfx_gl.dll!spvtools::opt::Pass::ProcessEntryPointCallTree(class std::function<bool > &,class spvtools::ir::Module *)	C++
 	vfx_gl.dll!spvtools::opt::AggressiveDCEPass::ProcessImpl(void)	C++
 	vfx_gl.dll!spvtools::opt::PassManager::Run(class spvtools::ir::Module *)	C++
 	vfx_gl.dll!spvtools::Optimizer::Run(unsigned int const *,unsigned __int64,class std::vector<unsigned int,class std::allocator<unsigned int> > *)	C++
 	vfx_gl.dll!CVfxShaderValidatorGLslang::GetBinaryShader(unsigned char * * ppBinaryOut, unsigned int * pBinarySizeOut) Line 856	C++
 	vfx_gl.dll!CVfxGL::CompileShaderAndGetVariables(unsigned __int64 nStaticComboIndex, unsigned __int64 nDynamicComboIndex, const CVfx * pVfx, VfxCompileTarget_t compileTarget, VfxProgram_t programType, VfxCompiledShaderInfo_t * pCompiledShaderInfoOut, bool bUseShaderCache, unsigned int nVfxCompileShaderFlags) Line 2861	C++
 	vfxcompile.dll!Worker_OnIndividualShader(IVfx * pVfxInterface, CVfx * pShaderGeneralHeader, unsigned __int64 nStaticCombo, unsigned __int64 nDynamicCombo, VfxCompileTarget_t compileTarget, VfxProgram_t programType, CVfxByteCodeManager & byteCodeManager, CShaderStats & shaderStats) Line 361	C++
 	vfxcompile.dll!Worker_ProcessWorkUnitFn(int iThread, unsigned __int64 nWorkUnit, MessageBuffer * pBuf) Line 596	C++
 	vfxcompile.dll!Local_ProcessWorkUnitFn(int iThread, int iWorkUnit) Line 736	C++
 	vfxcompile.dll!ThreadWorkerFunction(int iThread, void * pUserData) Line 85	C++
 	vfxcompile.dll!InternalRunThreadsFn(void * pParameter) Line 170	C++
 	tier0.dll!ThreadProcWrapped(unsigned __int64(*)(void *) pFunc, void * pParam) Line 535	C++
 	tier0.dll!ThreadProcConvert(void * pParam) Line 588	C++
```
This can reproduced from the command-line using the following command:

```
spirv-opt.exe --inline-entry-points-exhaustive --convert-local-access-chains --eliminate-local-single-block --eliminate-local-single-store --eliminate-insert-extract --eliminate-dead-code-aggressive --eliminate-dead-branches --merge-blocks --eliminate-local-single-block --eliminate-local-single-store --eliminate-local-multi-store --eliminate-insert-extract --eliminate-dead-code-aggressive --eliminate-common-uniform -o test0024_opt.spv test0024.spv
```

The previous known good in glslang did not have this crash.  I suspect due to the location of the crash that it may be related to PR #863.

[test0024.zip](https://github.com/KhronosGroup/SPIRV-Tools/files/1402898/test0024.zip)

**Comment by greg-lunarg** [comment]
Claiming it.

**Comment by dneto0** [comment]
Thanks.  I'll watch for a PR.

**Comment by greg-lunarg** [comment]
PR #883 fixes this problem. I have pushed a fix for the only problem that was noted to this point.

I know this PR fixes what ultimately will be dead code, but I would rather switch over to the cfg cleanup code in a more orderly manner and not in answer to a bug.

**Comment by danginsburg** [comment]
If possible, please merge PR #883 to master.  @greg-lunarg believes he has addressed the concerns from code review and this issue is currently causing a crash in our shader compilation workload so is gating us from taking the latest glslang.

**Comment by greg-lunarg** [comment]
@danginsburg An alternative to waiting for a merge of #883 is to move up to 5834719fc17d4735fce0102738b87b70255cfd5f and add CreateCFGCleanupPass after CreateDeadBranchElimPass. In general, you can leave this in your recipe as it will soon be added to the general recipe. I have a PR in to glslang that makes this commit the latest known good and adds CreateCFGCleanupPass to the recipe for legalization and optimization.

**Comment by dneto0** [comment]
Fix has been pushed into master as e3a720933046
Sorry for the delay.

I'll close this but please re-open if the same issue arises.

**Comment by danginsburg** [comment]
Confirmed fixed.  Thanks all.
